rllib/models/base_model.py

```diff
@@ -0,0 +1,244 @@
+"""Model base classes.
+
+RLlib models all inherit from the recurrent base class, which can be chained together
+with other models.
+
+The models input and output TensorDicts. Which keys the models read/write to
+and the desired tensor shapes must be defined in input_spec, output_spec,
+prev_state_spec, and next_state_spec.
+
+The `unroll` function gets the model inputs and previous recurrent state, and outputs
+the model outputs and next recurrent state. Note all ins/outs must match the specs.
+Users should override `_unroll` rather than `unroll`.
+
+`initial_state` returns the "next" state for the first recurrent iteration. Again,
+users should override `_initial_state` instead.
+
+For non-recurrent models, users may instead override `_forward` which does not
+make use of recurrent states.
+"""
+
+
+import abc
+from typing import Optional, Tuple, Any
+
+
+# TODO: Remove once TensorDict is in master
+class TensorDict:
+    def __init__(self, *args, **kwargs):
+        self.args = args
+        self.kwargs = kwargs
+
+    def filter(self, specs):
+        return {k for k in self.kwargs if k in specs.kwargs}
+
+    def __eq__(self, other: "TensorDict"):
+        return True
+
+
+# TODO: Remove once SpecDict is in master
+class SpecDict:
+    def __init__(self, *args, **kwargs):
+        self.args = args
+        self.kwargs = kwargs
+        pass
+
+    def validate(self, spec: Any) -> bool:
+        return True
+
+
+# TODO: Remove once ModelConfig is in master
+class ModelConfig:
+    pass
+
+
+ForwardOutputType = TensorDict
+# [Output, Recurrent State(s)]
+UnrollOutputType = Tuple[TensorDict, TensorDict]
+
+
+class RecurrentModel(abc.ABC):
+    """The base model all other models are based on."""
+
+    def __init__(self, name: Optional[str] = None):
+        self._name = name or self.__class__.__name__
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    # TODO: Should these be properties or functions? How do we know the
+    # input_spec before init? Won't our model change shapes depending
+    # on arguments passed to init?
```

Comment:
I think properties are better. With functions you are allowing the users of this abstract class to define their own set of input arguments which may not be result in subjective use of this abstraction and may not work with other components that we'll create down the line. I don't understand your comment about the knowing `input_spec` before the init? You can just use the attributes to code up the content of `input_spec`. Look at the Encoder example in here: https://github.com/kouroshHakha/ray/blob/rllib_2.0_policy_api/rllib2/models/torch/encoder.py#L33